scipy/signal/spectral.py

```diff
@@ -267,121 +267,495 @@ def welch(x, fs=1.0, window='hanning', nperseg=256, noverlap=None, nfft=None,
+    elif mode == 'magnitude':
+        result = np.absolute(result)
+    elif mode == 'angle' or mode == 'phase':
+        result = np.angle(result)
+    elif mode == 'complex':
+        pass
+
+    result *= scale
+    if mode == 'psd' and sides == 'onesided':
+        result[...,1:-1] *= 2
+
+    t = np.arange(nfft//2, len(x) - nfft//2 + 1, nfft - noverlap)/fs
+
+    if sides != 'twosided' and not nperseg % 2:
+        # get the last value correctly, it is negative otherwise
+        freqs[-1] *= -1
+
+    # we unwrap the phase here to handle the onesided vs. twosided case
+    if mode == 'phase':
+        result = np.unwrap(result, axis=-1)
+
+    result = result.astype(outdtype)
+
+    # All imaginary parts are zero anyways
+    if same_data:
+        result = result.real
 
     if axis != -1:
-        Pxx = np.rollaxis(Pxx, -1, axis)
+        result = np.rollaxis(result, -1, axis)
+
+    return freqs, result, t
 
-    return f, Pxx
+
+def _fft_helper(x, win, detrend_func, nperseg, noverlap, nfft):
+    '''
+    Calculate windowed FFT, for internal use by scipy.signal._spectral_helper
+
+    This is a helper function that does the main FFT calculation for 
+    _spectral helper. All input valdiation is performed there, and the data 
+    axis is assumed to be the last axis of x. It is not designed to be called
+    externally. The windows are not averaged over; the result from each window
+    is returned.
+
+    Returns
+    -------
+    result : ndarray
+        Array of FFT data
+
+    References
+    ----------
+    stackoverflow: Repeat NumPy array without replicating data?
+    <http://stackoverflow.com/a/5568169>
+
+    Notes
+    -----
+    Adapted from matplotlib.mlab
+    '''
+    # Created strided array of data segments
+    if nperseg == 1 and noverlap == 0:
+        result = x[..., np.newaxis]
+    else:
+        step = nperseg - noverlap
+        shape = x.shape[:-1]+((x.shape[-1]-noverlap)//step, nperseg)
+        strides = x.strides[:-1]+(step*x.strides[-1], x.strides[-1])
+        result = np.lib.stride_tricks.as_strided(x, shape=shape, 
+                                                 strides=strides)
+
+    # Detrend each data segment individually
+    result = detrend_func(result)
+    
+    # Create strided array of window profiles, same shape as result
+    if result.shape[-1] == 1:
+        winrep = win
+    else:
+        win_strides = np.zeros(len(result.shape), dtype='int64')
+        win_strides[-1] = win.strides[0]
+        winrep = np.lib.stride_tricks.as_strided(win, shape=result.shape, 
+                                                 strides=win_strides)
+
+    # Apply window by multiplication
+    result = winrep * result
+
+    # Zero pad here, now that windowing is done
+    pad_shape = list(result.shape)
+    pad_shape[-1] = nfft - pad_shape[-1]
+    result = np.concatenate((result, np.zeros(pad_shape)), axis=-1)
+
+    # Perform the fft
+    result = fftpack.fft(result, n=nfft, axis=-1)
```

Comment:
`fftpack.fft` will do the zero padding for you and defaults to operating along the last axis.  All the code to do zero padding can go away just above, and this could (but doesn't have to) become `result = fftpack.fft(result, n=nfft)`.